src/merkle.rs

```diff
@@ -18,49 +25,109 @@ fn hash_internal_node(left: &Hash, right: &Hash) -> Hash {
-        let mut tree = Merkle { tree: vec![] };
-        tree.build(data);
+        let leaves: Vec<Hash> = data.iter().map(|element| hash(element)).collect();
+        let mut tree = MerkleTree {
+            tree: vec![],
+            leaves,
+        };
+        tree.build();
         tree
     }
 
     /// Builds the Merkle Tree using a recursive bottom-up approach.
-    fn build(&mut self, data: &[Vec<u8>]) {
-        // Initialize tree with leaf-level hashes
-        let mut current_level: Vec<Hash> = data.iter().map(|element| hash(element)).collect();
-        let mut tree = current_level.clone();
-
-        // Build tree bottom-up
-        let mut count_leaves = current_level.len();
-        while count_leaves > 1 {
-            let mut level_hashes: Vec<Hash> = vec![];
-            for i in (0..count_leaves).step_by(2) {
-                if i + 1 < count_leaves {
-                    // Combine adjacent hashes
-                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i + 1]));
+    fn build(&mut self) {
+        let mut levels: Vec<Vec<Hash>> = Vec::new();
+        // Add the leaves at first level
+        levels.push(self.leaves.clone());
+
+        // Build subsequent levels until we reach the root
+        let mut current_level = self.leaves.clone();
+        while current_level.len() > 1 {
+            let mut next_level: Vec<Hash> = Vec::new();
+
+            // process pair of nodes
+            for i in (0..current_level.len()).step_by(2) {
+                let left = current_level[i];
+                let right = if i + 1 < current_level.len() {
+                    current_level[i + 1]
                 } else {
-                    // Clone the last hash to create a pair
-                    level_hashes.push(hash_internal_node(&current_level[i], &current_level[i]));
-                }
+                    left
+                };
+
+                let parent = hash_internal_node(&left, &right);
+                next_level.push(parent);
             }
+            levels.push(next_level.clone());
+            current_level = next_level;
+        }
+        self.tree = levels;
+    }
+
+    /// Returns the root hash of the Merkle tree.
+    pub fn root(&self) -> Hash {
+        *self.tree.last().unwrap().first().unwrap()
+    }
 
-            // Update current level and extend tree
-            current_level = level_hashes.clone();
-            tree.extend(level_hashes);
-            count_leaves = (count_leaves + 1) / 2;
+    // A proof is a list of hashes that can be used to verify the membership of a leaf in the tree
+    // For now the proof is simply that a  list of hashes.
+    // Possible improvements:
+    // 1. Store the direction of the hash (left or right) and the level of the tree
+    pub fn generate_proof(&self, leaf_index: usize) -> Option<MerkleProof> {
+        // Implement this for refactored Merkle Tree
+        if leaf_index >= self.leaves.len() {
+            return None;
```

Comment:
I think this should be a Result rather than an option. I think option type is mostly used for things that could "be nothing".

I think in this case, if you try to generate a proof for a node that does not exist, that is conceptually "an error".

It's pretty subjective however, but I think I'd make more sense.